BossMod/QuestBattle/QuestBattle.cs

```diff
@@ -0,0 +1,257 @@
+﻿using Dalamud.Game.ClientState.Conditions;
+
+namespace BossMod.QuestBattle;
+
+public record struct Waypoint(Vector3 Position, bool Pathfind = true);
+
+public enum NavigationStrategy
+{
+    // keep moving to destination even if in combat
+    Continue,
+    // clear waypoints upon entering combat
+    Stop,
+    // clear waypoints upon entering combat, repath when combat ends
+    Pause,
+}
+
+public class QuestObjective(WorldState ws)
+{
+    public readonly WorldState World = ws;
+    public string Name { get; private set; } = "";
+    public readonly List<Waypoint> Connections = [];
+    public NavigationStrategy NavigationStrategy = NavigationStrategy.Pause;
+
+    public string DisplayName
+    {
+        get
+        {
+            if (Name.Length > 0)
+                return Name;
+
+            if (Connections.Count > 0)
+                return Utils.Vec3String(Connections.Last().Position);
+
+            return "<none>";
+        }
+    }
+
+    public bool ForceStopNavigation;
+    public bool Completed;
+
+    public Action<Actor, AIHints> AddAIHints = (_, _) => { };
+    public Action<Actor> OnModelStateChanged = (_) => { };
+    public Action<Actor, ActorStatus> OnStatusGain = (_, _) => { };
+    public Action<Actor, ActorStatus> OnStatusLose = (_, _) => { };
+    public Action<Actor> OnActorEventStateChanged = (_) => { };
+    public Action<Actor, ushort> OnEventObjectStateChanged = (_, _) => { };
+    public Action<Actor, ushort, ushort> OnEventObjectAnimation = (_, _, _) => { };
+    public Action<Actor> OnActorCreated = (_) => { };
+    public Action<Actor> OnActorDestroyed = (_) => { };
+    public Action<Actor> OnActorCombatChanged = (_) => { };
+    public Action<Actor> OnActorKilled = (_) => { };
+    public Action<Actor> OnActorTargetableChanged = (_) => { };
+    public Action<WorldState.OpDirectorUpdate> OnDirectorUpdate = (_) => { };
+    public Action<ConditionFlag, bool> OnConditionChange = (_, _) => { };
+    public Action OnNavigationComplete = () => { };
+    public Action Update = () => { };
```

Comment:
Is there a reason to define empty functors? The rest of the codebase uses the following pattern:
```
Action<whatever>? MyAction;
MyAction += callbacks;
MyAction?.Invoke(args);
```

Reply:
Oh i didn't know you could use `+=` on nullable action

Reply:
fixed